### PR TITLE
Add apache HTTP request stats

### PR DESCRIPTION
**Pull request by dagnir**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have read the **CONTRIBUTING** document
- [ ] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have read the **README** document
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/ApacheHttpClient.java [hunk 1]
```diff
@@ -206,11 +214,15 @@ private boolean isProxyEnabled(ProxyConfiguration proxyConfiguration) {
...
     @Override
     public ExecutableHttpRequest prepareRequest(HttpExecuteRequest request) {
+        final MetricCollector metricCollector = request.metricCollector().orElseGet(NoOpMetricCollector::create);
+        metricCollector.reportMetric(HTTP_CLIENT_NAME, clientName());
         HttpRequestBase apacheRequest = toApacheRequest(request);
         return new ExecutableHttpRequest() {
             @Override
             public HttpExecuteResponse call() throws IOException {
-                return execute(apacheRequest);
+                final HttpExecuteResponse executeResponse = execute(apacheRequest);
```
Comment (zoewangg) [inline http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/ApacheHttpClient.java:223]: nit, no `final` on local variable 😛

File: http-client-spi/src/main/java/software/amazon/awssdk/http/HttpMetrics.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+    /**
+     * The number of connections from the connection pool that are busy serving requests.
+     */
+    public static final SdkMetric<Integer> LEASED_CONNECTIONS = metric("LeasedConnections", Integer.class);
+
+    /**
+     * The number of requests awaiting a free connection from the pool.
+     */
+    public static final SdkMetric<Integer> PENDING_REQUESTS = metric("PendingRequests", Integer.class);
```
Comment (zoewangg) [inline http-client-spi/src/main/java/software/amazon/awssdk/http/HttpMetrics.java:48]: `PENDING_CONNECTION_REQUESTS` or `PENDING_CONNECTIONS`?

File: http-client-spi/src/main/java/software/amazon/awssdk/http/HttpMetrics.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+public final class HttpMetrics {
+    /**
+     * The name of the HTTP client.
+     */
+    public static final SdkMetric<String> HTTP_CLIENT_NAME = metric("HttpClientName", String.class);
+
+    /**
+     * The maximum number of connections that will be pooled by the HTTP client.
+     */
+    public static final SdkMetric<Integer> MAX_CONNECTIONS = metric("MaxConnections", Integer.class);
```
Comment (zoewangg) [inline http-client-spi/src/main/java/software/amazon/awssdk/http/HttpMetrics.java:33]: Do we need to namespace all sdk metrics?

Comment (zoewangg) [inline http-client-spi/src/main/java/software/amazon/awssdk/http/HttpMetrics.java:33]: Do we want to differentiate h2 max connection from it? Is it worth creating a new class for all h2 http metrics such as window size, max streams, etc?

File: http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/ApacheHttpClient.java (hunk 1 above)
Comment (dagnir) [reply http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/ApacheHttpClient.java:223]: Ooops, this was Intellij's fault!

File: http-client-spi/src/main/java/software/amazon/awssdk/http/HttpMetrics.java (hunk 2 above)
Comment (dagnir) [reply http-client-spi/src/main/java/software/amazon/awssdk/http/HttpMetrics.java:48]: Yeah I'm not married to these names. Howe about `PENDING_CONNECTION_ACQUIRES`?

File: http-client-spi/src/main/java/software/amazon/awssdk/http/HttpMetrics.java (hunk 3 above)
Comment (dagnir) [reply http-client-spi/src/main/java/software/amazon/awssdk/http/HttpMetrics.java:33]: I think the issue with that is if you want to "pretty print" the metrics, or write your own publisher, there's no easy way to just get the metric name

Comment (dagnir) [reply http-client-spi/src/main/java/software/amazon/awssdk/http/HttpMetrics.java:33]: Yeah I think we should create additional metrics for H2 specific things; perhaps per-connection metrics?

**Comment by dagnir** [comment]
> Can we add some functional tests using the fake protocol test clients 

Can't do this yet unfortunately until all the other things are hooked up in the request pipeline and codegen so that an SDK client actually creates a collector and passes it to the HTTP client

**Pull request merged**